src/main/scala/beam/utils/StuckFinder.scala

```diff
@@ -0,0 +1,55 @@
+package beam.utils
+
+import beam.agentsim.scheduler.BeamAgentScheduler.ScheduledTrigger
+import beam.sim.config.BeamConfig.Beam.Debug.StuckAgentDetection
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/** THIS CLASS IS NOT THREAD-SAFE!!! It's safe to use it inside actor, but not use the reference in Future and other threads..
+  */
+class StuckFinder(val stuckAgentDetectionCfg: StuckAgentDetection) extends LazyLogging {
+  private val helper = new StuckFinderHelper[ScheduledTrigger]
+
+
+  def add(time: Long, st: ScheduledTrigger): Unit = {
+    helper.add(time, st)
+  }
+
+  def removeByKey(st: ScheduledTrigger): Option[ValueWithTime[ScheduledTrigger]] = {
+    helper.removeByKey(st)
+  }
+
+  def detectStuckAgents(time: Long = System.currentTimeMillis()): Seq[ValueWithTime[ScheduledTrigger]] = {
+    @tailrec
+    def detectStuckAgents0(stuckAgents: mutable.ArrayBuffer[ValueWithTime[ScheduledTrigger]]): Seq[ValueWithTime[ScheduledTrigger]] = {
+      helper.removeOldest match {
+        case Some(oldest) =>
+          val isStuck: Boolean = isStuckAgent(oldest.value, oldest.time, time)
```

Comment:
The changes applied still don't address my comment of pushing this into the helper itself. If you do it there then it seems that you could avoid thrashing (removing and re-adding).